### PR TITLE
feat/prune graph

### PR DESCRIPTION
**Pull request by admons**
- [ ] Ready for review
- [ ] Follows [CONTRIBUTING](https://github.com/snyk/dep-graph/blob/master/.github/CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

### CODE REVIEW

File: src/core/prune-graph.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+type NodeId = string;
+
+export function pruneGraph(baseDepGraph: DepGraph): DepGraph {
+  const depGraph = baseDepGraph as DepGraphInternal;
+
+  const depGraphBuilder = new DepGraphBuilder(
+    depGraph.pkgManager,
+    depGraph.rootPkg,
+    depGraph.rootNodeId,
```
Comment (admons) [inline src/core/prune-graph.ts:12]: remove support for different rootNodeId

File: src/core/dep-graph.ts [hunk 2]
```diff
@@ -327,26 +303,41 @@ class DepGraphImpl implements types.DepGraphInternal {
...
-      const out = this.pathsFromNodeToRoot(id).map((path) => {
-        return [this.getNodePkg(nodeId)].concat(path);
-      });
-      for (const path of out) {
-        allPaths.push(path);
-      }
-    });
+    for (const id of parentNodesIds) {
+      allPaths.push(
+        ...this.pathsFromNodeToRoot(id, ancestors.concat(nodeId)).map((path) =>
```
Comment (darscan) [inline src/core/dep-graph.ts:324]: Large spreads fail (hence not using spread previously)

Comment (darscan) [reply src/core/dep-graph.ts:324]: https://github.com/snyk/dep-graph/commit/0bd6afbfdec6131ea7ca55357930b1edec736d2a

File: src/core/dep-graph.ts [hunk 3]
```diff
@@ -327,26 +303,41 @@ class DepGraphImpl implements types.DepGraphInternal {
...
 
     return allPaths;
   }
 
-  private countNodePathsToRoot(nodeId: string): number {
+  private countNodePathsToRoot(
+    nodeId: string,
+    ancestors: string[] = [],
+  ): number {
+    if (ancestors.includes(nodeId)) {
```
Comment (darscan) [inline src/core/dep-graph.ts:337]: Again, this is weird. Surely you want this before recursing into `countNodePathsToRoot` not afterwards

File: src/core/dep-graph.ts [hunk 4]
```diff
@@ -327,26 +303,41 @@ class DepGraphImpl implements types.DepGraphInternal {
     return node;
   }
 
-  private pathsFromNodeToRoot(nodeId: string): types.PkgInfo[][] {
+  private pathsFromNodeToRoot(
+    nodeId: string,
+    ancestors: string[] = [],
```
Comment (darscan) [inline src/core/dep-graph.ts:308]: Are you very sure you don't want this to be a Set? (isn't `Set.has` faster than `Array.includes` for large sets?)

**Pull request closed without merging**